bench/bm_buffer.cpp

```diff
@@ -0,0 +1,153 @@
+
+    // init producers
+    std::atomic<int> threadID = 0;
+    std::vector<std::thread> producers;
+    for (int i = 0; i < nProducer; i++) {
+        producers.emplace_back([&]() {
+            // set thread affinity
+            setCpuAffinity(threadID++);
+            barrier.arrive_and_wait();
+            for (int rep = 0; rep < n_repeat; ++rep) {
+                BufferWriter auto writer = buffer.new_writer();
+                std::size_t nSamplesProduced = 0;
+                barrier.arrive_and_wait();
+                while (nSamplesProduced <= (min_samples + nProducer - 1) / nProducer) {
+                    writer.publish([](auto &) {}, vector_length);
+                    nSamplesProduced += vector_length;
+                }
+                barrier.arrive_and_wait();
+            }
+        });
+    }
+
+    // init consumers
+    std::vector<std::thread> consumers;
+    for (int i = 0; i < nConsumer; i++) {
+        consumers.emplace_back([&]() {
+            setCpuAffinity(threadID++);
+            barrier.arrive_and_wait();
+            for (int rep = 0; rep < n_repeat; ++rep) {
+                BufferReader auto reader = buffer.new_reader();
+                std::size_t nSamplesConsumed = 0;
+                barrier.arrive_and_wait();
+                while (nSamplesConsumed < min_samples) {
+                    if (reader.available() < vector_length) {
+                        continue;
+                    }
+                    const auto &input = reader.get(vector_length);
+                    nSamplesConsumed += input.size();
+
+                    if (!reader.consume(input.size())) {
+                        throw std::runtime_error(fmt::format("could not consume {} samples", input.size()));
+                    }
+                }
+                barrier.arrive_and_wait();
+            }
+        });
+    }
+
+    // all producers and consumer are ready, waiting to give the sign
+    barrier.arrive_and_wait();
+    ::benchmark::benchmark<n_repeat>(fmt::format("{}: {} writers -<{}>-> {} readers", name, nProducer, vector_length,
+                                                 nConsumer),
+                                     min_samples)
+            = [&]() {
+                  barrier.arrive_and_wait();
+                  barrier.arrive_and_wait();
+              };
+
+    // clean up
+    for (std::thread &thread : producers) thread.join();
+    for (std::thread &thread : consumers) thread.join();
+}
+
+inline const boost::ut::suite _buffer_tests = [] {
+    const uint64_t samples = 100'000'000; // minimum number of samples
+
+    for (int nP = 1; nP <= 4; nP *= 2) {
+        for (int nR = 1; nR <= 4; nR *= 2) {
+            if (not (nP == 1 and nR == 1)) {
+                benchmark::results::add_separator();
+            }
+            for (int buffer_type = 4; buffer_type <= 5; ++buffer_type) {
+                for (int veclen = 16; veclen <= 4096; veclen *= 16) { // vector length (<-> size of bulk operation)
```

Comment:
I think we should limit/reduce the test to just one veclen (e.g. '1' and or '1024') since the test primarily only pushes the writer/reader atomic lock-guards to the limit and does not do real processing of the data. This is also seen by the fact that the results pretty much scale with factors of 16.

Reply:
Maybe we should also add separators only when the number of writers change? :thinking: 

Would it make sense to align the test case names so that the 'writers' and 'readers' in the text align vertically? This would imply adding some fixed number formatting width and/or spaces.

Reply:
Different `veclen` basically reduces the number of calls until wrap-around in the buffer happens and when the writer might have to block because of slow readers. But in this artificial benchmark I agree that it's unlikely to make a difference.
Before I added the sweep over `veclen` I added separators only for different number of writers, yes. The separator is first and foremost a reading aid. So grouping ~3-7 lines was my goal.

Reply:
Umm, let me take that back. `veclen` does make a difference, as noted above.

Reply:
Maybe an option is to change/reshuffle the order of the loops from

now:  'for 1 ... nProducer' -> 'for 1 ... nReader' -> 'for buffer_type' -> 'for veclen'

to:   'for buffer_type'  -> 'for veclen' -> 'for 1 ... nProducer' -> 'for 1 ... nReader'
and 
  * make the buffer_type not a number (<-> historic reasons) but an enum
  * make the veclen loop a loop over a sorted list starting with [1, 1024, ...]
the latter could be -- alongside 'nProducer' and 'nReader' -- eventually be defined by an 'argv[]' statement (not now, but one of the future PR, once we know how we want to handle cmd-line parameters).
What do you think?

Reply:
N.B. 1, 1024, ...  would be complemented by power-of-two. The intermediate sizes < 1024 do not have real domain use cases but the '1' is sort of a stress test pushing the race condition for the atomics the most. 

I'd recommend keeping vector length sizes above 1024 and larger writer/reader combinations optional because these become quickly opinionated, depend on the available HW (the initial assumption was that most have PCs with at least 8 cores/threads), and use case (e.g. latency-critical application favour veclen<1024, throughput-related application usually veclen>>1024).

Reply:
Hmm, also interesting could be different `veclen` for producer / consumer. In one case the consumer is faster and must often wait for the producer, in the other case the producer hits the last consumer and must block.
I'll reorder the loops in any case.

Reply:
There are probably many more variations. I'd propose keeping this KISS and maybe adding benchmark tests for other corner tests later?

Reply:
![image](https://user-images.githubusercontent.com/3306474/224646357-c5dcb8f9-e1cf-4e4c-9612-dd0bc1740fde.png)
Fine. I'll push this state in a different branch then.